core/capability-negotiation.md

```diff
@@ -0,0 +1,457 @@
+- `CAP REQ` to blindly request a particular set of capabilities.
+
+Following this, the client MUST send the standard `NICK` and `USER` IRC commands.
+
+Upon receiving either a `CAP LS` or `CAP REQ` command during connection registration, the
+server MUST not complete registration until the client sends a `CAP END` command to
+indicate that capability negotiation has ended. This allows clients to request their
+desired capabilities before completing registration.
+
+Once capability negotiation has completed with a client-send `CAP END` command, registration
+continues as normal.
+
+With the above registration process, clients will either recieve a `CAP` message indicating
+that registration is paused while capability negotiation happens, or registration will
+complete immediately, indicating that the server doesn't support capability negotiation.
+
+Example with capability negotiation:
+
+    Client: CAP LS 302
+    Client: NICK dan
+    Client: USER d * 0 :This is a really good name
+    Server: CAP * LS :multi-prefix sasl
+    Client: CAP REQ :multi-prefix
+    Server: CAP * ACK multi-prefix
+    Client: CAP END
+    Server: 001 dan :Welcome to the Internet Relay Network dan
+    ...
+
+Example with capability negotiation, but where the client recognises no advertised caps: 
+
+    Client: CAP LS 302
+    Client: NICK dan
+    Client: USER d * 0 :This is a really good name
+    Server: CAP * LS :draft/example-1 draft/example-2
+    Client: CAP END
+    Server: 001 dan :Welcome to the Internet Relay Network dan
+    ...
+
+Example where the server doesn't support capability negotiation:
+
+    Client: CAP LS 302
+    Client: NICK dan
+    Client: USER d * 0 :This is a really good name
+    Server: 001 dan :Welcome to the Internet Relay Network dan
+    ...
+
+Example where the client doesn't support capability negotiation:
+
+    Client: NICK dan
+    Client: USER d * 0 :This is a really good name
+    Server: 001 dan :Welcome to the Internet Relay Network dan
+    ...
+
+## The CAP command
+
+The client capability negotiation extension is implemented by the addition of one command
+with several subcommands.  The command added is named `CAP`.  `CAP` takes a single
+required subcommand, optionally followed by a single parameter of space-separated capability
+identifiers.  Each capability in the list may be preceded by a capability modifier.
+
+The subcommands for `CAP` are: `LS`, `LIST`, `REQ`, `ACK`, `NAK`, and `END`.
+
+The `LS`, `LIST`, `REQ`, `ACK` and `NAK` subcommands may be followed by a single parameter
+containing a space-separated list of capabilities.  If more than one capability is named,
+the RFC1459 designated sentinel (`:`) for a multi-parameter argument must be present. The
+list of capabilities MUST be parsed from left to right and capabilities SHOULD only be sent
+once per command. If a capability is sent multiple times, the last one received takes priority.
```

Comment:
Maybe move this to the CAP values section, as it doesn't really have any impact on valueless caps.

Reply:
hmm, I'm not sure how this relates to valueless caps. in general it seems strange to say that if a cap is sent multiple times, the last one takes priority anyway. why is this recommended?

if I was the client getting a list of caps from the server, the last one would be the one that is applied, but if I was a server getting, say, a `REQ` from the client then I'd act on the first one in the list...

we should make this much more explicit, instead of just shifting it around. what is this sentence actually trying to say to vendors?